protocols/mempool-overview.md

```diff
@@ -105,25 +105,30 @@ To this end, we first provide a (brief) definition of what is a ledger, that is
 At process $p$, the $i$-th entry of the ledger is denoted $p.ledger[i]$.
 This entry contains either a null value ($\bot$), or a set of transactions, aka., a block.
-The height of the ledger at $p$ is the index of the first null entry.
+The height of the ledger at $p$ is the index of the first null entry; denoted $p.height$.
 Operation $submit(txs, i)$ attempts to write the set of transactions $txs$ to the $i$-th entry of the ledger.
-As standard, the ledger ensures that there is no gap between two entries at each process,
-that is  $\forall i. \forall p. p.ledger[i] \neq \bot \implies (i=0 \vee p.ledger[i-1] \neq \bot)$.
-It also makes sure that no two correct processes have different ledger entries (agreement);
-formally: $\forall i. \forall p,q \in Correct. (p.ledger[i] = \bot) \vee (q.ledger[i] = \bot) \vee (p.ledger[i] = q.ledger[i])$.
-Finally, the ledger requires that if some transaction appears at an index $i$, then a process submitted it at that index (validity).
-All the transactions in the non-null entries of the ledger are denoted $p.committed$;
-formally $p.committed = \\{ tx : \exists j. tx \in p.ledger[j] \\}$.
-The (history) variable $p.submitted$ holds all the transactions submitted so far by $p$.
+The (history) variable $p.submitted[i]$ holds all the transactions (if any) submitted by $p$.
+By extension, $p.submitted$ are all the transaction submitted by $p$.
+A transaction is committed when it appears in some entries of the ledger.
+We write $p.committed$ the committed transactions at $p$.
+
+As standard, the ledger ensures that:  
+* _(Gap-freedom)_ There is no gap between two entries at a correct process:  
+$\forall i. \forall p \in Correct. \square(p.ledger[i] \neq \bot \implies (i=0 \vee p.ledger[i-1] \neq \bot))$;  
+* _(Agreement)_ No two correct processes have different ledger entries; formally:  
+$\forall i. \forall p,q \in Correct. \square((p.ledger[i] = \bot) \vee (q.ledger[i] = \bot) \vee (p.ledger[i] = q.ledger[i]))$;  
+* _(Validity)_ If some transaction appears at an index $i$ at a correct process, then a process submitted it at that index:  
+$\forall p \in Correct. \forall i. \square(tx \in p.ledger[i] \implies tx \in \bigcup_q q.submitted[i]$).
+* _(Termination)_ If a correct process submits a block at its current height, eventually its height get incremented:  
+$\forall p \in Correct. \square((h=p.height \wedge p.submitted[h] \neq \varnothing) \implies \lozenge(p.height>h))$  
 
 **Mempool.**
 A mempool is a replicated set of transactions.
 At a process $p$, we write it $p.mempool$.
-We also define $p.hmempool$, the (history) variable that contains all the transactions ever added to the mempool by process $p$.
-
+We also define $p.hmempool$, the (history) variable that tracks all the transactions ever added to the mempool by process $p$.
 Below, we list the invariants of the mempool (at a correct process).
 
 The mempool is used as an input for the ledger:  
```

Comment:
We need to be careful here. `PrepareProposal`, in its current form (as of `v0.37.x`) violates this property, as the Application can include TXs directly in the proposal (and those TXs aren't added to the mempool)

Reply:
Can we find a way to have this case represented here? Even if we don't consider `PrepareProposal`, the ledger (blockchain) contains information, that can be called a transaction or not, which was never been part of the mempool. For instance, it contains evidence of misbehavior, handled by the evidence protocol; it contains validator changes, consensus parameters changes, etc.

So, in a general way, we should allow for the ledger and blocks to contain transactions that were not retrieved from the mempool, even though we don't want to model more complex possibilities introduced with ABCI++.